LSP crash on _some_ modifications of a file imported in another

(yeah, convoluted title, but the repro is somewhat convoluted too, although I'm sure there could be a simpler one if I really understood the problem)

**Describe the bug**

If `bar.ncl` imports `baz.ncl`, and I modify `baz.ncl` in certain ways, then the LSP panics with a typechecking-related error.

See the gif below:

![recording](https://github.com/tweag/nickel/assets/7226587/b6037ad0-2a9e-4024-8e80-9e0161d9114d)

The LSP crashes the instant I type the backtick character.
Not every character triggers the crash. From some quick testing, it seems that at least the alphanumeric characters, `&`, `/` and `@` don't cause it (but everything else I've tried, including things that are part of the grammar like `'` cause it).

<details><summary>LSP log and backtrace</summary>

```
2024-06-05T17:35:26.450 helix_lsp::transport [INFO] nls -> {"jsonrpc":"2.0","method":"textDocument/didChange","params":{"contentChanges":[{"text":"{\n  '\n  x,\n}\n"}],"textDocument":{"uri":"file:///tmp/tmp.MTYXidOIhu/baz.ncl","version":18}}}
2024-06-05T17:35:26.450 helix_lsp::transport [ERROR] nls err <- "thread 'main' panicked at core/src/typecheck/mod.rs:2793:26:\n"
2024-06-05T17:35:26.450 helix_lsp::transport [ERROR] nls err <- "Internal error: resolved import not found during typechecking.\n"
2024-06-05T17:35:26.450 helix_lsp::transport [ERROR] nls err <- "stack backtrace:\n"
2024-06-05T17:35:26.450 helix_lsp::transport [INFO] nls <- {"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"diagnostics":[{"message":"unexpected token","range":{"end":{"character":3,"line":1},"start":{"character":2,"line":1}},"relatedInformation":[],"severity":1},{"message":"","range":{"end":{"character":3,"line":1},"start":{"character":2,"line":1}},"severity":4}],"uri":"file:///tmp/tmp.MTYXidOIhu/baz.ncl"}}
2024-06-05T17:35:26.454 helix_lsp::transport [ERROR] nls err <- "   0:     0x5583fe667f77 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h4886110ae7835fd2\n"
2024-06-05T17:35:26.454 helix_lsp::transport [ERROR] nls err <- "   1:     0x5583fe69f9e0 - core::fmt::write::h6fab83f0f5bc0c8f\n"
2024-06-05T17:35:26.454 helix_lsp::transport [ERROR] nls err <- "   2:     0x5583fe682075 - std::io::Write::write_fmt::h64824cb874054607\n"
2024-06-05T17:35:26.454 helix_lsp::transport [ERROR] nls err <- "   3:     0x5583fe667d44 - std::sys_common::backtrace::print::hb8fea9d041077138\n"
2024-06-05T17:35:26.454 helix_lsp::transport [ERROR] nls err <- "   4:     0x5583fe6688b7 - std::panicking::default_hook::{{closure}}::hc60f0f690c67943c\n"
2024-06-05T17:35:26.454 helix_lsp::transport [ERROR] nls err <- "   5:     0x5583fe668617 - std::panicking::default_hook::h56281bcc5d6e2f17\n"
2024-06-05T17:35:26.454 helix_lsp::transport [ERROR] nls err <- "   6:     0x5583fe668e28 - std::panicking::rust_panic_with_hook::hd61e30fe5971d187\n"
2024-06-05T17:35:26.454 helix_lsp::transport [ERROR] nls err <- "   7:     0x5583fe668d0a - std::panicking::begin_panic_handler::{{closure}}::h95d2684e28f89fd2\n"
2024-06-05T17:35:26.454 helix_lsp::transport [ERROR] nls err <- "   8:     0x5583fe6681a6 - std::sys_common::backtrace::__rust_end_short_backtrace::h48780d898316fff8\n"
2024-06-05T17:35:26.454 helix_lsp::transport [ERROR] nls err <- "   9:     0x5583fe668a54 - rust_begin_unwind\n"
2024-06-05T17:35:26.454 helix_lsp::transport [ERROR] nls err <- "  10:     0x5583fdfe69c5 - core::panicking::panic_fmt::hdb89b5c14a056cec\n"
2024-06-05T17:35:26.454 helix_lsp::transport [ERROR] nls err <- "  11:     0x5583fdfe7083 - core::option::expect_failed::h27bb97517ce6ecdf\n"
2024-06-05T17:35:26.454 helix_lsp::transport [ERROR] nls err <- "  12:     0x5583fe312b48 - nickel_lang_core::typecheck::apparent_type::apparent_type_check_cycle::h99ed9efde0589b4c\n"
2024-06-05T17:35:26.454 helix_lsp::transport [ERROR] nls err <- "  13:     0x5583fe311ba4 - nickel_lang_core::typecheck::binding_type::h99fb2226ace2bef7\n"
2024-06-05T17:35:26.454 helix_lsp::transport [ERROR] nls err <- "  14:     0x5583fe11a0a3 - nickel_lang_core::typecheck::walk::h018d5afc526ad349\n"
2024-06-05T17:35:26.454 helix_lsp::transport [ERROR] nls err <- "  15:     0x5583fe1162cb - nickel_lang_core::typecheck::type_check_with_visitor::h80d2621bc539c33a\n"
2024-06-05T17:35:26.454 helix_lsp::transport [ERROR] nls err <- "  16:     0x5583fe0bdf2a - <nickel_lang_core::cache::Cache as nls::cache::CacheExt>::typecheck_with_analysis::h0990007f53c1dc77\n"
2024-06-05T17:35:26.454 helix_lsp::transport [ERROR] nls err <- "  17:     0x5583fe04685f - nls::world::World::typecheck::h0cd585e708bef479\n"
2024-06-05T17:35:26.454 helix_lsp::transport [ERROR] nls err <- "  18:     0x5583fe0f93d1 - nls::files::handle_save::hf94b8957bcbf2471\n"
2024-06-05T17:35:26.454 helix_lsp::transport [ERROR] nls err <- "  19:     0x5583fe05925f - nls::server::Server::handle_message::h70ce56c31ba13d67\n"
2024-06-05T17:35:26.454 helix_lsp::transport [ERROR] nls err <- "  20:     0x5583fe057f2d - nls::server::Server::run::h9b4e1a179adb8dc2\n"
2024-06-05T17:35:26.454 helix_lsp::transport [ERROR] nls err <- "  21:     0x5583fe10da37 - nls::main::h2b0835bd328a43af\n"
2024-06-05T17:35:26.455 helix_lsp::transport [ERROR] nls err <- "  22:     0x5583fe103583 - std::sys_common::backtrace::__rust_begin_short_backtrace::h66567cb7208ae91d\n"
2024-06-05T17:35:26.455 helix_lsp::transport [ERROR] nls err <- "  23:     0x5583fe10359d - std::rt::lang_start::{{closure}}::hbe3c71ccfbaad6dc\n"
2024-06-05T17:35:26.455 helix_lsp::transport [ERROR] nls err <- "  24:     0x5583fe668954 - std::panicking::try::ha86251cf5daa9bea\n"
2024-06-05T17:35:26.455 helix_lsp::transport [ERROR] nls err <- "  25:     0x5583fe6580db - std::rt::lang_start_internal::ha6a51778162f8d22\n"
2024-06-05T17:35:26.455 helix_lsp::transport [ERROR] nls err <- "  26:     0x5583fe10f0e5 - main\n"
2024-06-05T17:35:26.455 helix_lsp::transport [ERROR] nls err <- "  27:     0x7fc9ac4e710e - __libc_start_call_main\n"
2024-06-05T17:35:26.455 helix_lsp::transport [ERROR] nls err <- "  28:     0x7fc9ac4e71c9 - __libc_start_main@@GLIBC_2.34\n"
2024-06-05T17:35:26.455 helix_lsp::transport [ERROR] nls err <- "  29:     0x5583fdfe7345 - _start\n"
2024-06-05T17:35:26.455 helix_lsp::transport [ERROR] nls err <- "  30:                0x0 - <unknown>\n"
2024-06-05T17:35:26.461 helix_lsp::transport [ERROR] nls err: <- StreamClosed
```

</details>

**To Reproduce**

1. Create the files:
   ```sh
    cat <<EOF > bar.ncl
    let baz = import "baz.ncl" in
    {}
    EOF
    cat <<EOF > baz.ncl
    {
      x,
    }
    EOF
   ```
2. Open them in an editor with LSP properly setup
   ```sh
   hx -v bar.ncl baz.ncl --hsplit
   ```
3. Add a backtick before the `x` in `baz.ncl`

**Expected behavior**
A clear and concise description of what you expected to happen.

**Environment**
 - OS name + version:
 - Version of the code: 1.6.0

**Additional context**


That looks like it could be related to https://github.com/tweag/nickel/issues/1935 . Not 100% sure though

--- a/core/src/cache.rs
+++ b/core/src/cache.rs
@@ -1013,6 +1013,13 @@ impl Cache {
             .map(|TermEntry { state, .. }| std::mem::replace(state, new))
     }
 
+    /// Remove the cached term associated with this id.
+    ///
+    /// The file contents associated with this id remain, and they will be re-parsed if necessary.
+    pub fn reset(&mut self, file_id: FileId) {
+        self.terms.remove(&file_id);
+    }
+
     /// Retrieve the state of an entry. Return `None` if the entry is not in the term cache,
     /// meaning that the content of the source has been loaded but has not been parsed yet.
     pub fn entry_state(&self, file_id: FileId) -> Option<EntryState> {

--- a/core/src/cache.rs
+++ b/core/src/cache.rs
@@ -1013,11 +1013,36 @@ impl Cache {
             .map(|TermEntry { state, .. }| std::mem::replace(state, new))
     }
 
-    /// Remove the cached term associated with this id.
+    /// Remove the cached term associated with this id, and any cached terms
+    /// that import it.
     ///
-    /// The file contents associated with this id remain, and they will be re-parsed if necessary.
-    pub fn reset(&mut self, file_id: FileId) {
-        self.terms.remove(&file_id);
+    /// The file contents associated with this id remain, and they will be
+    /// re-parsed if necessary.
+    ///
+    /// This invalidation scheme is probably too aggressive; there are
+    /// situations where a change in one file doesn't require invalidation
+    /// of other files that import it. For example, if the parse status (i.e.
+    /// success/failure) of a file doesn't change, files that import it don't
+    /// need to re-resolve their imports. If the checked type of a file doesn't
+    /// change, files that import it don't need to be re-typechecked.
+    ///
+    /// Returns all the additional (i.e. not including the passed one) file ids
+    /// whose caches were invalidated.
+    pub fn invalidate_cache(&mut self, file_id: FileId) -> Vec<FileId> {
+        fn invalidate_rec(slf: &mut Cache, acc: &mut Vec<FileId>, file_id: FileId) {
+            slf.terms.remove(&file_id);
+            slf.imports.remove(&file_id);
+            let rev_deps = slf.rev_imports.remove(&file_id).unwrap_or_default();
+
+            acc.extend(rev_deps.iter().copied());
+            for f in &rev_deps {
+                invalidate_rec(slf, acc, *f);
+            }
+        }
+
+        let mut ret = vec![];
+        invalidate_rec(self, &mut ret, file_id);
+        ret
     }
 
     /// Retrieve the state of an entry. Return `None` if the entry is not in the term cache,

--- a/lsp/nls/src/files.rs
+++ b/lsp/nls/src/files.rs
@@ -23,7 +23,8 @@ pub(crate) fn uri_to_path(uri: &Url) -> std::result::Result<PathBuf, Error> {
     }
 }
 
-pub fn handle_open(server: &mut Server, params: DidOpenTextDocumentParams) -> Result<()> {
+/// Returns a list of open files that were potentially invalidated by the changes.
+pub fn handle_open(server: &mut Server, params: DidOpenTextDocumentParams) -> Result<Vec<Url>> {
     let id: RequestId = format!(
         "{}#{}",
         params.text_document.uri, params.text_document.version
@@ -49,10 +50,11 @@ pub fn handle_open(server: &mut Server, params: DidOpenTextDocumentParams) -> Re
         server.issue_diagnostics(*rev_dep, diags);
     }
     Trace::reply(id);
-    Ok(())
+    Ok(server.world.uris(invalid).cloned().collect())
 }
 
-pub fn handle_save(server: &mut Server, params: DidChangeTextDocumentParams) -> Result<()> {
+/// Returns a list of open files that were potentially invalidated by the changes.
+pub fn handle_save(server: &mut Server, params: DidChangeTextDocumentParams) -> Result<Vec<Url>> {
     let id: RequestId = format!(
         "{}#{}",
         params.text_document.uri, params.text_document.version
@@ -80,5 +82,5 @@ pub fn handle_save(server: &mut Server, params: DidChangeTextDocumentParams) -> 
         server.issue_diagnostics(*f, errors);
     }
     Trace::reply(id);
-    Ok(())
+    Ok(server.world.uris(invalid).cloned().collect())
 }

--- a/lsp/nls/src/files.rs
+++ b/lsp/nls/src/files.rs
@@ -76,7 +76,7 @@ pub fn handle_save(server: &mut Server, params: DidChangeTextDocumentParams) -> 
     server.issue_diagnostics(file_id, diags);
 
     for f in &invalid {
-        let errors = server.world.typecheck(*f).err().unwrap_or_default();
+        let errors = server.world.parse_and_typecheck(*f);
         server.issue_diagnostics(*f, errors);
     }
     Trace::reply(id);

--- a/lsp/nls/src/server.rs
+++ b/lsp/nls/src/server.rs
@@ -183,10 +183,13 @@ impl Server {
                     serde_json::from_value::<DidOpenTextDocumentParams>(notification.params)?;
                 let uri = params.text_document.uri.clone();
                 let contents = params.text_document.text.clone();
-                crate::files::handle_open(self, params)?;
+                let invalid = crate::files::handle_open(self, params)?;
                 self.background_jobs
                     .update_file(uri.clone(), contents, &self.world);
                 self.background_jobs.eval_file(uri);
+                for uri in invalid {
+                    self.background_jobs.eval_file(uri);
+                }
                 Ok(())
             }
             DidChangeTextDocument::METHOD => {
@@ -195,10 +198,13 @@ impl Server {
                     serde_json::from_value::<DidChangeTextDocumentParams>(notification.params)?;
                 let uri = params.text_document.uri.clone();
                 let contents = params.content_changes[0].text.clone();
-                crate::files::handle_save(self, params)?;
+                let invalid = crate::files::handle_save(self, params)?;
                 self.background_jobs
                     .update_file(uri.clone(), contents, &self.world);
                 self.background_jobs.eval_file(uri);
+                for uri in invalid {
+                    self.background_jobs.eval_file(uri);
+                }
                 Ok(())
             }
             _ => Ok(()),

--- a/lsp/nls/src/world.rs
+++ b/lsp/nls/src/world.rs
@@ -85,7 +85,7 @@ impl World {
 
         // Invalidate the cache of every file that tried, but failed, to import a file
         // with a name like this.
-        let mut invalid = path
+        let failed_to_import = path
             .file_name()
             .and_then(|name| self.failed_imports.remove(name))
             .unwrap_or_default();
@@ -94,19 +94,15 @@ impl World {
         // cache if it was imported by an already-open file.
         let file_id = self.cache.replace_string(SourcePath::Path(path), contents);
 
-        // Invalidate any cached inputs that imported the newly-opened file, so that any
-        // cross-file references are updated.
-        invalid.extend(self.cache.get_rev_imports_transitive(file_id));
+        let mut invalid = failed_to_import.clone();
+        invalid.extend(self.cache.invalidate_cache(file_id));
+
+        for f in failed_to_import {
+            invalid.extend(self.cache.invalidate_cache(f));
+        }
 
         for rev_dep in &invalid {
             self.analysis.remove(*rev_dep);
-            // Reset the cached state so that it will re-resolve its imports.
-            //
-            // Note that this will cause the contents to be re-parsed, which might be avoidable
-            // in some situations. It's safest to completely reset the state because post-parsing
-            // transformations (especially import resolution) can change the cached term in ways
-            // that are invalidated by the changes we just received.
-            self.cache.reset(*rev_dep);
         }
 
         self.file_uris.insert(file_id, uri);
@@ -121,15 +117,13 @@ impl World {
         &mut self,
         uri: Url,
         contents: String,
-    ) -> anyhow::Result<(FileId, HashSet<FileId>)> {
+    ) -> anyhow::Result<(FileId, Vec<FileId>)> {
         let path = uri_to_path(&uri)?;
         let file_id = self.cache.replace_string(SourcePath::Path(path), contents);
 
-        let invalid = self.cache.get_rev_imports_transitive(file_id);
+        let invalid = self.cache.invalidate_cache(file_id);
         for f in &invalid {
             self.analysis.remove(*f);
-            // Reset the cached state so that it will re-resolve its imports.
-            self.cache.reset(*f);
         }
         Ok((file_id, invalid))
     }

--- a/lsp/nls/src/world.rs
+++ b/lsp/nls/src/world.rs
@@ -99,10 +99,13 @@ impl World {
 
         for rev_dep in &invalid {
             self.analysis.remove(*rev_dep);
-            // Reset the cached state (Parsed is the earliest one) so that it will
-            // re-resolve its imports.
-            self.cache
-                .update_state(*rev_dep, nickel_lang_core::cache::EntryState::Parsed);
+            // Reset the cached state so that it will re-resolve its imports.
+            //
+            // Note that this will cause the contents to be re-parsed, which might be avoidable
+            // in some situations. It's safest to completely reset the state because post-parsing
+            // transformations (especially import resolution) can change the cached term in ways
+            // that are invalidated by the changes we just received.
+            self.cache.reset(*rev_dep);
         }
 
         self.file_uris.insert(file_id, uri);
@@ -124,6 +127,8 @@ impl World {
         let invalid = self.cache.get_rev_imports_transitive(file_id);
         for f in &invalid {
             self.analysis.remove(*f);
+            // Reset the cached state so that it will re-resolve its imports.
+            self.cache.reset(*f);
         }
         Ok((file_id, invalid))
     }

--- a/lsp/nls/src/world.rs
+++ b/lsp/nls/src/world.rs
@@ -5,6 +5,7 @@ use std::{
 };
 
 use codespan::FileId;
+use log::warn;
 use lsp_server::{ErrorCode, ResponseError};
 use lsp_types::Url;
 use nickel_lang_core::{
@@ -355,5 +356,14 @@ impl World {
             }
         }
         inner(self, span).unwrap_or_default()
+    }
+
+    pub fn uris(&self, ids: impl IntoIterator<Item = FileId>) -> impl Iterator<Item = &Url> {
+        ids.into_iter().filter_map(|id| {
+            self.file_uris.get(&id).or_else(|| {
+                warn!("no uri for {id:?}");
+                None
+            })
+        })
     }
 }

--- a/lsp/nls/src/world.rs
+++ b/lsp/nls/src/world.rs
@@ -94,6 +94,8 @@ impl World {
         // cache if it was imported by an already-open file.
         let file_id = self.cache.replace_string(SourcePath::Path(path), contents);
 
+        // The cache automatically invalidates reverse-dependencies; we also need
+        // to track them, so that we can clear our own analysis.
         let mut invalid = failed_to_import.clone();
         invalid.extend(self.cache.invalidate_cache(file_id));
 

--- a/lsp/nls/tests/main.rs
+++ b/lsp/nls/tests/main.rs
@@ -63,3 +63,51 @@ fn refresh_missing_imports() {
         }
     }
 }
+
+// Regression test for #1943, in which the cache was not properly reset
+// for imports that changed state from parseable to not-parseable.
+#[test]
+fn reload_broken_imports() {
+    let _ = env_logger::try_init();
+    let mut harness = TestHarness::new();
+
+    let url = |s: &str| lsp_types::Url::from_file_path(s).unwrap();
+    harness.send_file(url("/dep.ncl"), "{ x }");
+
+    harness.send_file(url("/test.ncl"), "import \"dep.ncl\"");
+    let diags = harness.wait_for_diagnostics();
+
+    assert_eq!("/dep.ncl", diags.uri.path());
+    assert!(diags.diagnostics.is_empty());
+
+    let diags = harness.wait_for_diagnostics();
+    assert_eq!("/test.ncl", diags.uri.path());
+    assert!(diags.diagnostics.is_empty());
+
+    // We expect two more diagnostics coming from background eval.
+    let _diags = harness.wait_for_diagnostics();
+    let _diags = harness.wait_for_diagnostics();
+
+    // Introduce an error in the import.
+    harness.send_file(url("/dep.ncl"), "{ `x = 1 }");
+
+    // Check that we get back clean diagnostics for both files.
+    // (LSP doesn't define the order, but we happen to know it)
+    // Loop because we can get back the diagnostics twice from each
+    // file (once from synchronous typechecking, once from eval in the background).
+    loop {
+        let diags = harness.wait_for_diagnostics();
+        if diags.uri.path() == "/dep.ncl" {
+            assert_eq!(diags.diagnostics[0].message, "unexpected token");
+            break;
+        }
+    }
+
+    loop {
+        let diags = harness.wait_for_diagnostics();
+        if diags.uri.path() == "/test.ncl" {
+            assert_eq!(diags.diagnostics[0].message, "unexpected token");
+            break;
+        }
+    }
+}

